src/Geometry/Primitives/NonCSG/ConvexPolygon.jl

```diff
@@ -0,0 +1,104 @@
+# MIT license
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# See LICENSE in the project root for full license information.
+
+import Statistics
+import GeometricalPredicates
+
+"""
+    ConvexPolygon{T} <: Surface{T}
+
+General Convex Polygon surface, not a valid CSG object.
+The rotation of the polygon around its normal is defined by `rotationvec`.
+`rotationvec×surfacenormal` is taken as the vector along the u axis.
+
+```julia
+ConvexPolygon(local_frame::Transform{T}, local_polygon_points::Vector{Vector{T}}, interface::NullOrFresnel{T} = nullinterface(T))
+```
+
+The local frame defines the plane (spans by the right and up vectors) with the plane normal given by the forward vector.
+the local_polygon_points are given with respect to the local frame and are 2D points.
+"""
+struct ConvexPolygon{T} <: Surface{T}
+    plane::Plane{T,3}
+    local_frame::Transform{T}
+    local_points::Vector{Vector{T}}
+
+    # for efficency
+    _poly2d::GeometricalPredicates.Polygon2D{GeometricalPredicates.Point2D}
+
+    function ConvexPolygon(
+            local_frame::Transform{T},
+            local_polygon_points::Vector{Vector{T}}, 
+            interface::NullOrFresnel{T} = NullInterface(T)
+        ) where {T<:Real}
+
+        @assert length(local_polygon_points) > 3         # need at least 3 points to define apolygon
+        @assert size(local_polygon_points[1])[1] == 2    # check that first point is a 2D point
```

Comment:
if you use SVector instead of Vector then you can specify vector size variable declaration and can eliminate this assert.